### ISSUE TITLE
Support for toggling like slideToggle / fadeToggle

### ISSUE THREAD

**Issue by shaicoleman** [julianshapiro/velocity#276]
At the moment there's no equivalent to jQuery's slideToggle/fadeToggle functions, which makes migration from jQuery more difficult, as with velocity it requires to convert everything to multiple lines of codes

**Comment by rodrigocfd** [comment]
Right now I'm on a situation where slideToggle would be useful. I have a button which shows/hides another div, and this is the code I'm having to use:
`$obj.velocity($obj.is(':visible') ? 'slideUp' : 'slideDown', { duration:200 });`
Instead of just:
`$obj.velocity('slideToggle', { duration:200 });`

**Comment by mu3** [comment]
Since I use own shorthand `$.vel` anyway I added this little workaround 😸

    $.fn.vel = function(){
      if(arguments[0] === 'fadeToggle')
        arguments[0] = this.is(':visible') ? 'fadeOut' : 'fadeIn';
      return this.velocity.apply(this,arguments)
    };
Though apparently it doesn't work good with `stop` etc.

**Comment by unabst** [comment]
> There are very few cases where UI logic isn't dealing with additional in/out logic at the same time; the UI should almost always know which direction to go in.

This is a valid point, but there are cases when the direction is already internalized. Say a panel that shows or hides itself. A button that toggles the panel removes the abstractions of both "show" and "hide" from the logic of that code. All we want is to toggle two obvious states that can be beneath us, which are also visibly obvious in their consequences.

Now, maybe we want a button to be hot or cold depending on the panel being visible or hidden. We might try and synchronize the states by reading them, for which we would need to "know the direction". Or, if we have a simple button toggle function, we can just toggle both and internalize both the button and panel state. They just need to be synced once in the beginning. 

Basically, if there are buttons that toggle things, we need code that toggles things. It's also easy to implement so no big deal, but if you ever feel like revisiting toggle in the future ;)

**Comment by Rycochet** [comment]
Reopening as a reminder - got an idea related to this (but needs the v2 velocity to be ready first) :-)

**Comment by kkthxby3** [comment]
Adding support for this...

My use case is ScrollMagic with Velocity...  They don't work too well together.  Using Velocity on a scene does nothing when using ScrollMagic's 'duration' option.  Easy enough to work around using triggers.

However, animations do not reverse.  Using Velocity's 'fadeOut' at a particular trigger works fine.  But scrolling back in reverse it does not fade back in.  Haven't been able to find any way to get this to work using the libraries themselves.  Finally resorted to writing a function call at the triggers to have Velocity manually fade in/out.  And that is where the toggle would be nice.

The 'reverse' option requires something to have had first happened.  So basically I have to pull the element to check it's opacity each time the trigger is triggered...  Not a big deal but not that elegant either.

**Comment by Rycochet** [comment]
I think one of the biggest issues with things like this is the `display` type options - my idea (when I can get time for it) is to allow string based values such as `display` and `visibility` to be used as part of the normal properties map - using with force-feeding to begin with then makes `"reverse"` far more trivial to implement for toggles. This would need some easings that are basically "first frame change" or "last frame change", choosing automatically but allowing people to override if needed.

Current code:
``` javascript
Velocity(el, {
   opacity: 0
}, {
   display: "none"
});
```

New idea:
``` javascript
Velocity(el, {
   opacity: 0
   display: "none"
});
```

The second one is pretty obviously going to be easier (well, hopefully obvious lol) for reversing as all the properties are in one place, and not using shortcut options where properties are actually getting addressed - which should hopefully make `toggle` actions a lot easier to implement and more resilient too.

**Comment by Rycochet** [comment]
Going to close this for now - the best way to handle this is to use a new Sequence, and put it on a named queue, then simply call "reverse" on that queue when needed.

String based properties are in (display and visibility have moved over obviously), so you can now tell it when to switch the value with the string type "at-start", "at-end", and "during" easings.
